src/Symfony/Component/Serializer/Encoder/JsonEncoder.php

```diff
@@ -23,10 +23,10 @@ class JsonEncoder implements EncoderInterface, DecoderInterface
     protected $encodingImpl;
     protected $decodingImpl;
 
-    public function __construct(JsonEncode $encodingImpl = null, JsonDecode $decodingImpl = null)
+    public function __construct(array $defaultContext = [], JsonEncode $encodingImpl = null, JsonDecode $decodingImpl = null)
```

Comment:
This is a BC break, could you move this parameter at the end of the list please?

Reply:
I understand, but I'm not sure how to set different `defaultContext` for encoding and decoding, because `JsonEncode` and `JsonDecode` are not passed into `JsonEncoder` as a service, and other places are the same `defaultContext` set in `framework.serializer.default_context`, Like `XmlEncoder` or `DateTimeNormalizer`.

also I have tried providing the `defaultContext` parameter at the end, but this raises an error in `DefinitionErrorExceptionPass`: 

```
Invalid constructor argument 3 for service "serializer.encoder.json": argument 1 must be defined before. Check your service definition.
```

Do you have any idea?

Reply:
Actually, the proper solution is maybe to register them as services indeed!